### PR TITLE
[Messenger] Use "warning" instead of "error" log level for RecoverableException

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

When one throw a `RecoverableExceptionInterface`, one knows somethings
wrong occurs, so if an `error` should be emmited, it's in place where the
exception is thrown.

And, by extension, if nothing really wrong occurs (like a random network
error), one may want to retry the message **without** any errors.

That's why in this PR, if an exception of type `RecoverableExceptionInterface`
is thrown, it uses `warning` insteand of `error` log level. So
everything stay under the radar.

### CODE REVIEW

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php [hunk 1]
```diff
@@ -70,7 +70,17 @@ public function onMessageFailed(WorkerMessageFailedEvent $event)
...
 
             if (null !== $this->logger) {
-                $this->logger->error('Error thrown while handling message {class}. Sending for retry #{retryCount} using {delay} ms delay. Error: "{error}"', $context + ['retryCount' => $retryCount, 'delay' => $delay, 'error' => $throwable->getMessage(), 'exception' => $throwable]);
+                $logLevel = 'error';
+                if ($throwable instanceof RecoverableExceptionInterface) {
+                    $logLevel = 'warning';
+                } elseif ($throwable instanceof HandlerFailedException) {
+                    foreach ($throwable->getNestedExceptions() as $nestedException) {
+                        if ($nestedException instanceof RecoverableExceptionInterface) {
+                            $logLevel = 'warning';
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php]: `break;`

**Review by jderusse — APPROVED** [review]

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php (hunk 1 above)
Comment (lyrixx) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php]: Good catch, fixed 👍🏼

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php [hunk 2]
```diff
@@ -70,7 +70,18 @@ public function onMessageFailed(WorkerMessageFailedEvent $event)
             $delay = $retryStrategy->getWaitingTime($envelope, $throwable);
 
             if (null !== $this->logger) {
-                $this->logger->error('Error thrown while handling message {class}. Sending for retry #{retryCount} using {delay} ms delay. Error: "{error}"', $context + ['retryCount' => $retryCount, 'delay' => $delay, 'error' => $throwable->getMessage(), 'exception' => $throwable]);
+                $logLevel = 'error';
```
Comment (stof) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php]: should we use the `LogLevel` constants from psr/log here ?

Comment (lyrixx) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php]: Done 👍🏼

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php [hunk 3]
```diff
@@ -70,7 +71,18 @@ public function onMessageFailed(WorkerMessageFailedEvent $event)
...
 
             if (null !== $this->logger) {
-                $this->logger->error('Error thrown while handling message {class}. Sending for retry #{retryCount} using {delay} ms delay. Error: "{error}"', $context + ['retryCount' => $retryCount, 'delay' => $delay, 'error' => $throwable->getMessage(), 'exception' => $throwable]);
+                $logLevel = LogLevel::ERROR;
+                if ($throwable instanceof RecoverableExceptionInterface) {
+                    $logLevel = LogLevel::WARNING;
+                } elseif ($throwable instanceof HandlerFailedException) {
+                    foreach ($throwable->getNestedExceptions() as $nestedException) {
+                        if ($nestedException instanceof RecoverableExceptionInterface) {
+                            $logLevel = LogLevel::WARNING;
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php:80]: It should only be changed to warning if all of the handlers threw recoverable exceptions, no?

Comment (lyrixx) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php:80]: I copied the behavior from the shouldRetry method, So I think this is fine like this. But I'm not sure indeed

Comment (jderusse) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php:80]: Agreed: one handler throwing a `RecoverableExceptionInterface` is enougth to retry the message.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Pull request merged**